text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,568 @@
+So if we can find a solution that allows maintainers to move forward, helping
+users more on the edge, while not impacting users on older rust version, would be
+a big help.
+
+The sooner we improve the status quo, the better, as it can take years for
+these changes to percolate out to those exclusively developing with an older
+Rust version (in contrast with the example above).
+This delay can be reduced somewhat if a newer development version can be used
+without upgrading the MSRV.
+
+In solving this, we need to keep in mind
+- Users need to be aware when they are on old versions for evaluating security risk and when debugging issues
+- We don't want to end up like other ecosystems where no one can use new features
+  because users are stuck on 3-20 year old versions of the language specification.
+  The compatibility story is fairly strong with Rust, helping us keep
+  compiler and dependency upgrades cheap.
+- Some people keep their development and production MSRVs the same while others keep them separate, like with a `Cargo.msrv.lock`
+- A `Cargo.lock` should not resolve differently when upgrading Rust without any other action.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The `rust-version` field
+
+*(update to [manifest documentation](https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field))*
+
+The `rust-version` field is an optional key that tells cargo what version of the
+Rust language and compiler your package can be compiled with. If the currently
+selected version of the Rust compiler is older than the stated version, cargo
+will exit with an error, telling the user what version is required.
+To support this, Cargo will prefer dependencies that are compatible with your `rust-version`.
+
+The first version of Cargo that supports this field was released with Rust 1.56.0.
+In older releases, the field will be ignored, and Cargo will display a warning.
+
+```toml
+[package]
+# ...
+rust-version = "1.56"
+```
+
+The Rust version must be a bare version number with two or three components; it
+cannot include semver operators or pre-release identifiers. Compiler pre-release
+identifiers such as -nightly will be ignored while checking the Rust version.
+The `rust-version` must be equal to or newer than the version that first
+introduced the configured `edition`.
+
+The `rust-version` may be ignored using the `--ignore-rust-version` option.
+
+Setting the `rust-version` key in `[package]` will affect all targets/crates in
+the package, including test suites, benchmarks, binaries, examples, etc.
+
+## Rust Version
+
+*(update to [Dependency Resolution's Other Constraints documentation](https://doc.rust-lang.org/cargo/reference/resolver.html))*
+
+When multiple versions of a dependency satisfy all version requirements,
+cargo will prefer those with a compatible `package.rust-version` over those that
+aren't compatible.
```

Comment:
```suggestion
aren't compatible.
If this results in selecting a version other than the latest compatible version, Cargo
will attempt to issue a warning indicating that the package's rust-version prevented
selecting the newer version.
```

Reply:
I used "attempt" deliberately here. I realize that Cargo's resolver has a *limited* ability to detect such things. But I think it's worth trying, to the extent we can, so that the use of older version doesn't pass unnoticed.

Reply:
This is in the Guide section, specifically in proposed documentation updates for the Cargo book.  I don't think this clarification helps in that form of documentation and is unclear from a policy setting perspective.

As for the need being addressed, I am currently handling this in the `cargo update` section.  Knowing whether something was held back from MSRV and justifies a warning is hard to determine.  There is also the problem of versions being held back due to semver incompatibility.  My thought was `cargo update` could provide a view similar to `cargo upgrade` so you can see when dependencies are stale and what gaps exist between `Cargo.lock` and the latest on crates.io.  This makes it so we aren't having to guess intent but just serving information.

Reply:
>  Knowing whether something was held back from MSRV and justifies a warning is hard to determine.

Worst case, I wonder if we could resolve a second time with "maximum" (using only the cached downloaded index without attempting to refresh it) and report if any of the versions differ. The resolver isn't *that* expensive.

Reply:
>  I'd just like the RFC to suggest that we'll make a best effort to notify the user that MSRV is making their dependencies outdated. That would make switching the default to the MSRV-aware resolver much safer.

As I mentioned, the RFC already includes a solution (exclusively for `cargo update`).  Are you feeling that is insufficient?   Alternatively, I could move how we'll solve this to Unresolved Questions, giving `cargo update` and the warning as two example ways we could solve it, and defer the answer to the implementation / stabilization process.

> Worst case, I wonder if we could resolve a second time with "maximum" (using only the cached downloaded index without attempting to refresh it) and report if any of the versions differ. The resolver isn't that expensive.

I feel like we already have a death by a thousand duck bites for "cargo script" performance so I'm always wanting to watch out for that.  My care about there is no-change resolution performance.

So one potential way
1. Resolve by default
2. resolve again with the original inputs to the first pass but without an MSRV set
3. diff the resolution trees, reporting any differences.  I'm assuming this would be depth-first. recursing only on branches that stayed the same.

Depending on performance, we either always do the re-resolve or first check if the resolve changed anything.  I'm unsure which is more performant as a resolve when nothing changed is likely to be fast.

I've not checked the code so I have some uncertainty on the performance impact of ensuring the inputs are around for two separate resolve passes.

Speaking of performance, I wonder if we could hash the inputs and put that into the lockfile and bypass the "no change" resolution process.

Reply:
@epage Sure, https://github.com/rust-lang/rfcs/pull/3537/files#diff-23b157d7104c7952ceddeefe57163ed969ac0aa16ae394c38f937d9f24d71462R294 is probably sufficient to make this clear.

Reply:
> Speaking of performance, I wonder if we could hash the inputs and put that into the lockfile and bypass the "no change" resolution process.

Yeah, I would expect the need for a warning to only come up when attempting to *change* the resolution.

Reply:
I've captured some notes on this in the Alternatives.